### PR TITLE
where_exp filter should filter posts

### PR DESCRIPTION
**Pull request by pathawks**
```liquid
{{ site.posts | where_exp: "posts", "posts.title == 'My Title'" }}
```

This just returns `site.posts`, it does not filter posts by title.

### CODE REVIEW

**Comment by parkr** [comment]
@pathawks If someone filed an issue for that here, we'd likely ask them to change it to `{{ site.posts | where: "title", "My Title" }}`. It's definitely a bug but also good to know that there's another way for this particular operation.

Is it all `==` operations? Could you `assert_equal EXPECTED_SITE, results.size` instead of comparing against `site.posts`? Is the result supposed to be empty?

**Comment by pathawks** [comment]
Ultimately, my use case is to get a list of all posts from the past 12 months. When tge date stuff wasn’t working, I decided to see if I could get the `where_exp` filter to work _at all_, and I could not.

I know there is at least one post who’s title is `Foo Bar`. I will take another look tonight to see if I can make the test a bit tighter.

**Comment by parkr** [comment]
Ok, thanks @pathawks 😄 

You could assert that (1) the result is the right size (i.e. 1) and (2) that the result contains the post you're looking for (find the document yourself using `site.posts.find {|p| p.title == "Foo Bar"}` and compare)

**Comment by pathawks** [comment]
Changes have been made :+1:

Test still fails, which is the point. The filter is _not_ filtering, but just returning all posts.

**Comment by jeromecoupe** [comment]
Same here with Jekyll 3.1.6

- This works fine: '{% assign potsByAuthor = site.posts | where: 'author', page.teamId %}`
- Whereas this just returns all the posts `{% assign potsByAuthor = site.posts | where_exp: 'item', 'item.author == page.teamId' %}` The size of the `potsByAuthor` is always the total number of posts
- Same if I try with a string: `{% assign potsByAuthor = site.posts | where_exp: 'item', 'item.author == "jeromecoupe"' %}`

**Comment by parkr** [comment]
Hm, tests aren't being triggered. Would you mind rebasing on `jekyll:master` to try to re-trigger the build (and resolve conflicts)?

**Comment by pathawks** [comment]
Not sure how the failing test could be related (and only with Ruby 2.3). Would you mind restarting that build to see if it was just a fluke?

**Comment by pathawks** [comment]
Thanks! Looks like all tests pass now, and no conflicts with **master**

**Comment by parkr** [comment]
@pathawks on mobile so can't double check, but do we actually test filtering `site.posts`? I bet drops behave strangely :stuck_out_tongue:

File: test/test_filters.rb [hunk 1]
```diff
@@ -594,6 +603,18 @@ def to_liquid
         assert_equal "b", results[1]["id"]
         assert_equal "d", results[2]["id"]
       end
+
+      should "filter posts" do
+        site = fixture_site.tap(&:read)
+        results = @filter.where_exp(site.posts, "obj", "obj.title == 'Foo Bar'")
+        assert_equal 1, results.length
+        assert_equal site.posts.find { |p| p.title == "Foo Bar" }, results.first
```
Comment (pathawks) [inline test/test_filters.rb]: > do we actually test filtering `site.posts`?

:+1:

File: test/test_filters.rb [hunk 2]
```diff
@@ -594,6 +603,18 @@ def to_liquid
         assert_equal "b", results[1]["id"]
         assert_equal "d", results[2]["id"]
       end
+
+      should "filter posts" do
+        site = fixture_site.tap(&:read)
+        results = @filter.where_exp(site.posts, "obj", "obj.title == 'Foo Bar'")
```
Comment (parkr) [inline test/test_filters.rb]: This should be `site.site_payload["site"]["posts"]` to mimic the Liquid context.

**Comment by pathawks** [comment]
Travis has given up on me?

**Comment by pathawks** [comment]
Failure is because [Travis cannot install activesupport](https://travis-ci.org/jekyll/jekyll/jobs/144208193#L465)

**Comment by parkr** [comment]
Looks like we need to depend on activesupport v4.x for Ruby 2.1 and below?

**Comment by Crunch09** [comment]
@parkr [Yes](https://github.com/rails/rails/blob/5-0-stable/activesupport/activesupport.gemspec#L10), would you be interested in a PR for this?

**Comment by parkr** [comment]
@Crunch09, Yes!

**Comment by parkr** [comment]
#5100 is merged, would you mind rebasing?

**Comment by pathawks** [comment]
-[x] Rebased

File: test/test_filters.rb [hunk 3]
```diff
@@ -594,6 +603,19 @@ def to_liquid
         assert_equal "b", results[1]["id"]
         assert_equal "d", results[2]["id"]
       end
+
+      should "filter posts" do
+        site = fixture_site.tap(&:read)
+        posts = site.site_payload["site"]["posts"]
+        results = @filter.where_exp(posts, "obj", "obj.title == 'Foo Bar'")
+        assert_equal 1, results.length
+        assert_equal site.posts.find { |p| p.title == "Foo Bar" }, results.first
```
Comment (parkr) [inline test/test_filters.rb]: Here's a curiosity: why isn't this comparing a Document with a DocumentDrop? Can you ensure `results.first` is, in fact, a DocumentDrop?

**Comment by parkr** [comment]
@pathawks ❤️

File: test/test_filters.rb (hunk 3 above)
Comment (pathawks) [inline test/test_filters.rb]: `results.first` is a `Document`, not a `DocumentDrop`.
So... what should I do about that?

Comment (parkr) [inline test/test_filters.rb]: Maybe the CollectionDrop should be recursively calling to_liquid on the documents it's accessing. When and how Liquid calls this method has always been a mystery.

**Comment by parkr** [comment]
@jekyllbot: merge +bug

**Comment by pathawks** [comment]
I will not `@mention` in commit messages.
I will not `@mention` in commit messages.
I will not `@mention` in commit messages.

**Comment by parkr** [comment]
@pathawks ❤️

**Pull request merged**